sdk/tests/deqp/modules/shared/glsBufferTestUtil.js

```diff
@@ -0,0 +1,1068 @@
+     * @param {number} numBytes
+     * @param {Uint8Array} bytes
+     */
+    glsBufferTestUtil.BufferSubDataWriter.prototype.writeNoTarget = function(buffer, offset, numBytes, bytes) {
+        this.write(buffer, offset, numBytes, bytes, gl.ARRAY_BUFFER);
+    };
+
+    /**
+     * @param {WebGLBuffer} buffer
+     * @param {number} offset
+     * @param {number} numBytes
+     * @param {Uint8Array} bytes
+     * @param {number} target
+     */
+    glsBufferTestUtil.BufferSubDataWriter.prototype.write = function(buffer, offset, numBytes, bytes, target) {
+        gl.bindBuffer(target, buffer);
+        gl.bufferSubData(target, offset, bytes);
+        gl.bindBuffer(target, null);
+    };
+
+    // Buffer verifier system.
+
+    /**
+     * @enum {number}
+     */
+    glsBufferTestUtil.VerifyType = {
+        AS_VERTEX_ARRAY: 0,
+        AS_INDEX_ARRAY: 1,
+        AS_UNIFORM_BUFFER: 2,
+        AS_PIXEL_UNPACK_BUFFER: 3,
+        BUFFER_READ_MAP: 4
+    };
+
+    /**
+     * @param {glsBufferTestUtil.VerifyType} verify
+     * @return {string}
+     */
+    glsBufferTestUtil.getVerifyTypeDescription = function(verify) {
+        /** @type {Array<string>} */ var s_desc =
+        [
+            'rendering as vertex data',
+            'rendering as index data',
+            'reading in shader as uniform buffer data',
+            'using as PBO and uploading to texture',
+            'reading back using glMapBufferRange()'
+        ];
+
+        return /** @type {string} */ (deUtil.getArrayElement(s_desc, verify));
+    };
+
+    /**
+     * @constructor
+     */
+    glsBufferTestUtil.BufferVerifierBase = function() {};
+
+    /**
+     * //Meant to be overriden
+     * @return {number}
+     */
+    glsBufferTestUtil.BufferVerifierBase.prototype.getMinSize = function() { throw new Error('Must be overriden'); };
+
+    /**
+     * //Meant to be overriden
+     * @return {number}
+     */
+    glsBufferTestUtil.BufferVerifierBase.prototype.getAlignment = function() { throw new Error('Must be overriden'); };
+
+    /**
+     * @param {WebGLBuffer} buffer
+     * @param {Uint8Array} reference
+     * @param {number} offset
+     * @param {number} numBytes
+     * @return {boolean}
+     */
+    glsBufferTestUtil.BufferVerifierBase.prototype.verifyNoTarget = function(buffer, reference, offset, numBytes) {
+        throw new Error('Must be overriden');
+    };
+
+    /**
+     * //Meant to be overriden
+     * @param {WebGLBuffer} buffer
+     * @param {Uint8Array} reference
+     * @param {number} offset
+     * @param {number} numBytes
+     * @param {number} targetHint
+     * @return {boolean}
+     */
+    glsBufferTestUtil.BufferVerifierBase.prototype.verify = function(buffer, reference, offset, numBytes, targetHint) {
+        //In WebGL 2, ELEMENT_ARRAY_BUFFER and TRANSFORM_FEEDBACK_BUFFER cannot be rebound to a different
+        //type of buffer, so, let's copy their data to an ARRAY BUFFER and pass that one instead to be verified.
```

Comment:
Actually the restriction is only on ELEMENT_ARRAY_BUFFERs -- buffer objects initially bound to that binding point can't be bound to other binding points. See https://www.khronos.org/registry/webgl/specs/latest/2.0/#BUFFER_OBJECT_BINDING . If you're finding other behavior (in particular, for TRANSFORM_FEEDBACK_BUFFER) that's probably a bug.